source/Irrlicht/CNullDriver.cpp

```diff
@@ -182,10 +181,48 @@ void main()
 		params.CompareEnable = false;
 		params.CompareFunc = asset::ISampler::ECO_ALWAYS;
 		params.AnisotropicFilter = 4u;
+        params.LodBias = 0.f;
+        params.MinLod = -1000.f;
+        params.MaxLod = 1000.f;
 		auto sampler = core::make_smart_refctd_ptr<asset::ICPUSampler>(params);
 		addBuiltInToCaches(sampler,"irr/builtin/samplers/default");
+
+        params.TextureWrapU = params.TextureWrapV = params.TextureWrapW = asset::ISampler::ETC_CLAMP_TO_BORDER;
+        sampler = core::make_smart_refctd_ptr<asset::ICPUSampler>(params);
+        addBuiltInToCaches(sampler, "irr/builtin/samplers/default_clamp_to_border");
 	}
 
+    //ds layouts
+    core::smart_refctd_ptr<asset::ICPUDescriptorSetLayout> defaultDs1Layout;
+    {
+        asset::ICPUDescriptorSetLayout::SBinding bnd;
+        bnd.count = 1u;
+        bnd.binding = 0u;
+        //maybe even ESS_ALL_GRAPHICS?
+        bnd.stageFlags = static_cast<asset::ICPUSpecializedShader::E_SHADER_STAGE>(asset::ICPUSpecializedShader::ESS_VERTEX | asset::ICPUSpecializedShader::ESS_FRAGMENT);
+        bnd.type = asset::EDT_UNIFORM_BUFFER;
+        defaultDs1Layout = core::make_smart_refctd_ptr<asset::ICPUDescriptorSetLayout>(&bnd, &bnd+1);
+        //it's intentionally added to cache later, see comments below, dont touch this order of insertions
+    }
+
+    //desc sets
+    {
+        auto ds1 = core::make_smart_refctd_ptr<asset::ICPUDescriptorSet>(core::smart_refctd_ptr<asset::ICPUDescriptorSetLayout>(defaultDs1Layout.get()));
+        {
+            auto desc = ds1->getDescriptors(0u).begin();
+            //size of 4x4 matrix (MVP), another 4x4 (MV or M, whatever space u prefer for lighting calculation), 3x3 matrix (normal matrix) and vec3 (cam pos) including std140 padding, in this order
+            constexpr size_t UBO_SZ = (16u+16u+12u+4u) * sizeof(float);
```

Comment:
actually make a struct with proper packing, padding and alignment that people could then use to fill out data for the shader, and take the `sizeof` that

Reply:
Alos MV/M and 3x3 normal matrices are 12 bytes (same as `core::matrix3x4SIMD`) not 16

Reply:
also because a `mat3` takes the space of a `mat4x3` you can stash the `vec3` camera position in the unused column of a `mat3/mat4x3` when in `row_major` layout (fun fact, you can get it directly in GLSL without transpose because GLSL works on columns)

Reply:
ok but the proper struct?

Reply:
see newest commits and `asset_utils.h` file, this is very outdated

Reply:
see newest commits and `asset_utils.h` file, this is very outdated